File: iterator/DTUIterator.php
<?php
/**
 * イテレータパターン：反復子
 * 繰り返しがあるときに、次があったときに返す役割を担ってくれるもの
 * コレクションクラス（配列）とかのときに、共通処理として取れる（色んな型や注文にも応えられる）
 */

interface DTUIterator
{
    // 次があるかチェックする抽象メソッド
    public function hasNext();
    
    // 次を返す抽象メソッド
    public function next();
}

File: java/oreoreiterator/DTUIterator.java
package oreoreiterator;

import oreorecollection.DTUCollection;
import dtumember.DTUMember;

/**
 * DTUIterator
 */
public class DTUIterator implements AreIterator {

    private DTUCollection collection;
    private int index = 0;

    /**
     * コンストラクタでDTU組織を数え上げの対象に指定
     */
    public DTUIterator(DTUCollection collection) {
        this.collection = collection;
    }

    /**
     * 次があるか判定する
     * 
     * @return boolean 
     */
    public boolean hasNext()
    {
        if (this.index < this.collection.getScale()) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * 次があったらオブジェクトを返す
     * 
     * @return 
     */
    public DTUMember next()
    {
        DTUMember member = this.collection.getMember(this.index);
        this.index++;
        return member;
    }
}